ssl/record/rec_layer_d1.c

```diff
@@ -120,14 +120,18 @@ void DTLS_RECORD_LAYER_set_write_sequence(RECORD_LAYER *rl, unsigned char *seq)
 static int dtls1_copy_record(SSL *s, pitem *item)
 {
     DTLS1_RECORD_DATA *rdata;
+    SSL3_BUFFER *rbuf;
 
     rdata = (DTLS1_RECORD_DATA *)item->data;
+    rbuf = &s->rlayer.rbuf;
 
-    SSL3_BUFFER_release(&s->rlayer.rbuf);
+    if (s->options & SSL_OP_CLEANSE_PLAINTEXT)
+        OPENSSL_cleanse(rbuf->buf, rbuf->len);
+    SSL3_BUFFER_release(rbuf);
 
     s->rlayer.packet = rdata->packet;
     s->rlayer.packet_length = rdata->packet_length;
-    memcpy(&s->rlayer.rbuf, &(rdata->rbuf), sizeof(SSL3_BUFFER));
+    memcpy(rbuf, &(rdata->rbuf), sizeof(SSL3_BUFFER));
```

Comment:
Here we're copying plaintext out of `rdata`, which comes from `item->data`. There are two callers of `dtls1_copy_record`, and in both cases `item->data` is immediately freed, so we probably also need to cleanse that. This function could be called for any type of record - not just application data. Which makes me wonder whether this option is supposed to cleanse *all* plaintext (e.g. including handshake/alert plaintext), or just application data plaintext. If the former then I think there are some spots that might have been missed  (see further comments below).

Reply:
will recheck and update

Reply:
I am trying to get around where DTLS maintains copies of records.
 Sof far i found three queues
* _unprocessed_rcds.q_: 
  I think this contains ciphertext records (no cleansing required)
* _processed_rcds.q_: I think this contains plaintext records, which can be either app data or handshake. 
  Cleansing is required only when we free it and there is still data left (_DTLS_RECORD_LAYER_clear_)
* _buffered_app_data.q_: Bufferd but not yet deliverd plaintext application data records. Must be cleansed in (DTLS_RECORD_LAYER_clear, in dtls1_read_bytes)

Is it correct that each of these three queues contain DTLS1_RECORD_DATA items where each of those convey it own 16K buffer for a single record?

To my understanding dtls1_copy_record takes an item from one of those queues and replaces s->rlayer.rbuf, with the queued item. The old record is freed (and cleansed) in this process?

Note however, as i understand it _dtls1_copy_record_, does not copy the record it self, in only copies metadata (pointers and stuff).

Keeping this in mind I think we can revert the cleanse modification in _dtls1_copy_record_, since it **only cleanses application data after it is delivered (and already cleansed) to the application**, but we need to add cleansing of leftofer items in _buffered_app_data_ and _processed_rcds_ _DTLS_RECORD_LAYER_clear_.

I am really not totaly sure about my analysis, please let me know if it is correct. I already have prepared a fixup containing the following changes:

* Add cleansing in _DTLS_RECORD_LAYER_clear_
* Revert cleansing in _dtls1_copy_record_

Reply:
> Is it correct that each of these three queues contain DTLS1_RECORD_DATA items where each of those convey it own 16K buffer for a single record?

Yes

Reply:
> The old record is freed (and cleansed) in this process?

It is not cleansed. However, I don't think it needs to be. If the old record contains plaintext application data then we'll only copy in a new record if we've already read that data - and if we've read it then (as a result of this PR) we will have already cleansed that application data.

> I am really not totaly sure about my analysis, please let me know if it is correct. I already have prepared a fixup containing the following changes:

I agree with your analysis